README.md
# GBA_tranportation

transportation analysis for GBA

## 时间线

- 19年4月2号 12 a.m.
  [南沙大桥](https://zh.wikipedia.org/wiki/%E5%8D%97%E6%B2%99%E5%A4%A7%E6%A1%A5)（虎门二桥）开通
- 20年5月5号 `虎门大桥振动`
  2020年5月5日14时许，虎门大桥出现较为明显的波浪形抖动现象，随后大桥双向车道及通航水域被封闭通行。综合相关桥梁专家的转述，虎门大桥出现波动原因，主要是因为在桥面设置了“水马”，改变了钢箱梁的气动外形，与虎门大桥多年来的超负荷运行无关。
- 20年5月15号 9 a.m. `虎门大桥恢复通车`
  广东省交通集团发布消息称，15日9时，虎门大桥恢复通车。[广州日报全媒体](http://www.xinhuanet.com/politics/2020-05/16/c_1125991891.htm)记者昨天现场走访看到，重新通车后的虎门大桥，路面平整，车流顺畅，悬索桥两侧采取了抑振措施，以保障大桥安全。目前，虎门大桥继续实施全天禁止货车及40座以上客车通行。
  
## 桥梁简介

![桥梁位置示意图](.fig/Road%20networks%20in%20the%20GBA%20and%20key%20bridges%20across%20the%20Pearl%20River%20Estuary.jpg)

- [虎门大桥](https://zh.wikipedia.org/wiki/%E8%99%8E%E9%97%A8%E5%A4%A7%E6%A1%A5)
  虎门大桥是一座位于中国广东省珠江三角洲中部、连接东莞市虎门镇和广州市南沙区的大型悬索桥，位于珠江出海口的虎门水道和蒲州水道之上，是珠江三角洲高速公路网的重要组成部分，也是连接京珠高速公路和广深高速公路的重要交通枢纽。虎门大桥及其连接线工程全长15.6公里，其中主桥全长4.6公里，桥面为双向六车道设计，于1997年6月9日建成通车。
  自虎门大桥通车后，成为香港、深圳、东莞、惠州等珠三角东岸地区往返中山、珠海、佛山（顺德等地）、江门等珠三角西岸地区的必经之路，随着珠三角城市发展，大桥通车后十年日均车流量已接近七万架次，远超设计负荷，车龙有时更伸延至广深高速公路及其他相连路段，轮候上桥时间达两小时，在节假日内地高速公路免路费的时段车龙更长，因此在坊间有了“英雄难过虎门关（桥）”的戏言.

- [南沙大桥](https://zh.wikipedia.org/wiki/%E5%8D%97%E6%B2%99%E5%A4%A7%E6%A1%A5)
  南沙大桥（规划及建设时期称为虎门二桥、虎门第二公路通道）是一座位于中国广东省珠江三角洲中部、横跨珠江的公路悬索桥，也是S6广龙高速的起始路段。本桥距上游黄埔大桥约20公里。南沙大桥跨越珠江狮子洋的大沙水道和坭洲水道，东连东莞市沙田镇，西通广州市南沙区。南沙大桥是珠江三角洲高速公路网的重要组成部分，是连接广州绕城高速公路和广深沿江高速公路的重要过江通道，也是继虎门大桥后又一条沟通珠江出海口两岸的大型跨江悬索桥。
  4月2日上午10点，虎门二桥通车仪式在虎门二桥坭洲水道桥桥面举行。历经十年筹划、五年“智造”的虎门二桥终于迎来通车时刻。`中午12点，虎门二桥正式开通运营，向社会车辆开放`。

- [港珠澳大桥](https://zh.wikipedia.org/zh/%E6%B8%AF%E7%8F%A0%E6%BE%B3%E5%A4%A7%E6%A9%8B)
  港珠澳大桥启用后，大幅缩短了来往香港、珠海、澳门三地的交通时间，行车时间仅需一小时左右，而且不需要再绕经虎门大桥；而港珠澳也会形成“一小时生活圈”，对粤港澳大湾区发展有象征意义。大桥的使用者进入香港后亦可经香港的快速公路，包括深港西部通道前往深圳。

- [深中通道](https://zh.wikipedia.org/zh-hans/%E6%B7%B1%E4%B8%AD%E9%80%9A%E9%81%93)

## 数据

### [API](https://lbs.amap.com/api/webservice/guide/api/direction#driving)

驾车路径规划 API 可以规划以小客车、轿车通勤出行的方案，并且返回通勤方案的数据

- 驾车规划的响应结果
  ![](.fig/respond_driving.png)

- 导航路段信息step
  ![](.fig/respond_step.png)

### 数据时间范围

特殊日期：2019年4月5号~7号，清明假期

- 原来JCP文章的数据 (8+5=12天)
  
    e:\Data\GBA\GBA_trajectory_info_190322.csv
    e:\Data\GBA\GBA_trajectory_info_190323.csv
    e:\Data\GBA\GBA_trajectory_info_190324.csv
    e:\Data\GBA\GBA_trajectory_info_190325.csv
    e:\Data\GBA\GBA_trajectory_info_190327.csv
    e:\Data\GBA\GBA_trajectory_info_190328.csv
    f:\6_数据\GBA\gba_0326.zip
    f:\6_数据\GBA\gba_0329.zip
    f:\6_数据\GBA\gba_0330.zip
    f:\6_数据\GBA\gba_0331.zip
    f:\6_数据\GBA\gba_0401.zip

- 虎门二桥开通（19年4月2号）(10天)

    f:\6_数据\GBA\gba_0402.zip
    f:\6_数据\GBA\gba_0403.zip
    f:\6_数据\GBA\gba_0404.zip
    f:\6_数据\GBA\gba_0405.zip
    f:\6_数据\GBA\gba_0407.zip
    f:\6_数据\GBA\gba_0408.zip
    f:\6_数据\GBA\gba_0409.zip
    f:\6_数据\GBA\gba_0410.zip
    f:\6_数据\GBA\gba_0411.zip
    f:\6_数据\GBA\gba_0412.zip

- 虎门大桥振动关闭（9天）

    z:\Data\GBA\output\GBA_step_200506.csv
    z:\Data\GBA\output\GBA_step_200507.csv
    z:\Data\GBA\output\GBA_step_200508.csv
    z:\Data\GBA\output\GBA_step_200509.csv
    z:\Data\GBA\output\GBA_step_200510.csv
    z:\Data\GBA\output\GBA_step_200511.csv
    z:\Data\GBA\output\GBA_step_200512.csv
    z:\Data\GBA\output\GBA_step_200513.csv
    z:\Data\GBA\output\GBA_step_200514.csv

- 虎门大桥恢复通车（[15号早上九点](https://baijiahao.baidu.com/s?id=1666723785340949565&wfr=spider&for=pc)，20天）

    z:\Data\GBA\output\GBA_step_200515.csv
    z:\Data\GBA\output\GBA_step_200516.csv
    z:\Data\GBA\output\GBA_step_200517.csv
    z:\Data\GBA\output\GBA_step_200518.csv
    z:\Data\GBA\output\GBA_step_200519.csv
    z:\Data\GBA\output\GBA_step_200520.csv
    z:\Data\GBA\output\GBA_step_200521.csv
    z:\Data\GBA\output\GBA_step_200522.csv
    z:\Data\GBA\output\GBA_step_200523.csv
    z:\Data\GBA\output\GBA_step_200524.csv
    z:\Data\GBA\output\GBA_step_200525.csv
    z:\Data\GBA\output\GBA_step_200526.csv
    z:\Data\GBA\output\GBA_step_200527.csv
    z:\Data\GBA\output\GBA_step_200528.csv
    z:\Data\GBA\output\GBA_step_200529.csv
    z:\Data\GBA\output\GBA_step_200530.csv
    z:\Data\GBA\output\GBA_step_200531.csv
    z:\Data\GBA\output\GBA_step_200601.csv
    z:\Data\GBA\output\GBA_step_200602.csv
    z:\Data\GBA\output\GBA_step_200603.csv

## 相关报道

[南沙大桥今天正式通车！双向八车道 设计车速100公里/小时！](https://baijiahao.baidu.com/s?id=1629698920228371979&wfr=spider&for=pc)

### 粤港澳大湾区发展规划纲要

To build a rapid transport network in the Greater Bay Area.

`To focus on connecting the Mainland with Hong Kong and Macao, as well as connecting the east and west banks of the Pearl River Estuary, build a rapid inter-city transport network mainly involving high-speed rails, inter-city railway links and high-grade motorways, and strive for reducing travelling time among major cities within the Greater Bay Area to one hour or less.` To draw up a development plan for inter-city railway links in the Greater Bay Area, improve the backbone railway networks of the Greater Bay Area, speed up the construction of inter-city railways, and progressively plan urban railway transit projects in major cities of the PRD. `To accelerate the construction of the Shenzhen-Zhongshan Bridge and the Second Humen Bridge crossriver link`. To adopt new models for clearance procedures with a view to making better use of the Guangzhou-Shenzhen-Hong Kong Express Rail Link and the Hong Kong-Zhuhai-Macao Bridge. To take forward the planning and construction of new boundary crossings such as the Liantang/Heung Yuen Wai Boundary Control Point, the new Guangdong-Macao border access (Qingmao Boundary Control Point), Hengqin Port (exploring the relocation of the Cotai Frontier Port of Macao) and West Kowloon Station of the Guangzhou-Shenzhen-Hong Kong Express Rail Link.
To strengthen Hong Kong’s and Macao’s transport connections with the Mainland, pursue the effective connection of various modes of transport including urban railway transit, develop a safe and efficient interchange system, enhance the handling capacity and clearance facilitation of control points in Guangdong, Hong Kong and Macao, and promote an efficient and convenient flow of people and goods.



changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

- [x] Period 1~2: lacking the `step` and `tmcs` shapfile

- [ ] 计算不同城市之间的距离变化情况
- [ ] 绘制kde变化叠加图

  ```
  # 全部变量的KDE分布图
  dist_cols = 6
  dist_rows = len(test_data.columns)
  plt.figure(figsize=(4*dist_cols, 4 * dist_rows))

  i = 1
  for col in test_data.columns:
      ax = plt.subplot(dist_rows, dist_cols, i)
      ax = sns.kdeplot(train_data[col], color='Red', shade=True)
      ax = sns.kdeplot(test_data[col], color='Blue', shade=True)
      ax.set_xlabel(col)
      ax.set_ylabel("Freanquency")
      ax.legend(['train', 'test'])

      i+=1
  plt.show()
  ```

## [1.0.04] - 2020-06-30

### Added

- `trip_crawler` to crawl data from the map provider.
- `data_process_step_spark` reorgnize, and move some funs from eda_step

## [1.0.03] - 2020-06-30

### Added

- Extract bridge congestion index from `step` csv or shapefile with Spark.

## [1.0.02] - 2020-06-01

### Added

- 线路热力图
- spark 参与计算，提升速度

## [1.0.01] - 2021-04-18

### Added

- build the project
- `PathSet` Class: To store the path between cities
- 预处理数据（path, steps）


src/eda_step.py
#%%
import os
import datetime
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

from spark_helper import spark
from pyspark.sql.types import StringType, IntegerType
import pyspark.sql.functions as F

from utils.classes import PathSet
from df_helper import save_to_geojson, df_pipline
from df_spark_helper import parser_tripID_info, split_period, convert_2_dateFormat

# data initial
path_set = PathSet(cache_folder='../cache', file_name='wkt_step')
gdf_shps = path_set.get_shapes_with_bridges_info()
fids_bridge = gdf_shps[~gdf_shps.bridge.isnull()].fid.unique()


fig_title_font = {'family' : 'Times New Roman',
        'weight' : 'normal',
        'size' : 12,
}

#%%


""" preprocess csv data """
def add_date_to_file(fn):
    """add date to file

    Args:
        fn (function): [description]
    """
    date_ = '20'+fn.split("_")[-1].split('.')[0]
    df = pd.read_csv(fn)
    df.loc[:, "date"] = int(date_)
    
    print(f"{fn} -> {fn.replace('db/step', 'steps_csv')}")
    df.to_csv( fn.replace('db/step', 'steps_csv') , index=False)
    
    return


def add_date_batch(folder='/home/pcl/Data/GBA/db/step', n_jobs=56):
    from multiprocessing import Pool
    s = datetime.datetime.now()

    fns = [os.path.join(folder, f) for f in os.listdir(folder)]
    fns.sort()

    pools = Pool(n_jobs)
    data  = pools.map_async(add_date_to_file, fns).get()
    pools.close()
    pools.join()

    return


# if __name__ == '__main__':
    #data = read_data_batch()
    # add_date_batch()
    # s = datetime.datetime.now()
    # print(datetime.datetime.now() -s)


#%%
""" pyspark funcs """

def check_dataframe():
    """Check df
    """
    print('stepID Top 10')
    spark.sql(" select * from step order by stepID desc limit 10 ").show()
    
    print('time distribution')
    spark.sql("""select t, count(road) from step group by t order by t""").show(100)
    
    # 查看最大的step数值， 是小于100的
    print('stepID distribution')
    spark.sql("""select stepID, count(road) from step group by stepID order by stepID""").show(100)
    
    print('date distribution')
    spark.sql("""select date, count(road) from step group by date order by date """).show(30)

    return 


def check_num_dist():
    """对比分析shp and csv folder 可得，两者的记录是相符合的, 除了 2020-05-23
    """
    a = spark.sql('select date, count(tripID) as count from step group by date order by date').toPandas()
    b = spark.sql('select date, count(tripID) as count from csv group by date order by date').toPandas()
    c = b.merge(a, on='date', how='left')
    c.loc[:, 'same'] = (c.count_y - c.count_x) == 0
    
    return c


def check_tripID_dist(df_2):
    """check the distribution of the length of tripID

    Args:
        df_2 ([dataframe]): [description]
    """
    # check the number of tripID in p1 and p2
    steps = df_2.filter(F.col('date') < '2020-05-01')\
                .withColumn('tripID_num', F.length(df_2.tripID.cast(StringType())) )
    steps.groupBy(steps.date, steps.tripID_num).count().sort(steps.date).show()

    # check the number of tripID in p3 and p4
    steps = df_2.filter(F.col('date') >= '2020-05-01')\
                .withColumn('tripID_num', F.length(df_2.tripID.cast(StringType())) )
    steps.groupBy(steps.date, steps.tripID_num).count().sort(steps.date).show()


def read_and_tranform_data():
    """read origin csv data and save them in the parquet format

    Returns:
        [Boolean]: If success, return True
    """
    # 读取初始数据，然后写出到parquet格式, 
    df_1 = spark.read.format('csv').load("/home/pcl/Data/GBA/steps_all.csv", header=True, inferSchema=True,dateFormat="yyyyMMdd")
    df_1 = parser_tripID_info(df_1)
    df_1 = split_period(df_1)
    df_1 = df_1.repartition('date')

    # save to parquet format
    s = datetime.datetime.now()
    df_1.write.format('parquet').mode('overwrite').save('/home/pcl/Data/GBA/db/spark/steps_shp.parquet')
    print(datetime.datetime.now() -s)

    df_2 = spark.read.format('csv').load("/home/pcl/Data/GBA/steps_csv", header=True, inferSchema=True, dateFormat="yyyyMMdd")
    df_2 = convert_2_dateFormat(df_2)
    # check_tripID_dist(df_2)

    s = datetime.datetime.now()
    df_2.repartition('date').write.format('parquet').mode('overwrite').save('/home/pcl/Data/GBA/db/spark/steps_csv.parquet')
    print(datetime.datetime.now() -s)
    
    return True


def cal_step_freq(df, verbose=False):
    """Count the step frequency of each OD in each period
    
    Return [pd.DataFrame]
    """
    
    view_table = 'step_'
    df.createOrReplaceTempView("step_")
    sql = f"""SELECT a.OD, a.period, a.fid, a.nums, a.nums/b.totoal_nums AS freq
            FROM 
            (
                SELECT OD, fid, period, count(tripID) AS nums
                FROM {view_table}
                GROUP BY OD, period, fid
            ) AS a, 
            (
                SELECT OD, period, count(DISTINCT(timestamp)) AS totoal_nums 
                FROM {view_table}
                GROUP BY OD,  period
            ) as b
            WHERE a.OD == b.OD and a.period == b.period
            ORDER BY a.OD, a.period, freq
        """

    if verbose: print(sql)
    res = spark.sql(sql)

    return res.toPandas()


def selct_data_by_date(date_ = '2020-05-18'):
    df = spark.sql(f"select * from step where date == '{date_}'").cache()
    df_ = spark.sql(f"select * from csv where date == '{date_}'").cache()
    df.show()
    df_.show()
    
    return df, df_


#%%
# TODO 空值的处理， road里边的null为空值

df = spark.read.format('parquet').load('/home/pcl/Data/GBA/db/spark/steps_p3_p4.parquet')
# df = spark.read.format('parquet').load('/home/pcl/Data/GBA/db/spark/steps_shp.parquet')
df.createOrReplaceTempView('step')

df_csv = spark.read.format('parquet').load('/home/pcl/Data/GBA/db/spark/steps_csv.parquet')
df_csv.createOrReplaceTempView('csv')


#%%

def find_the_freq_steps(df_csv):
    # TODO
    df_csv.show()
    total_num = df_csv.count()

    cols_drop = ['tripID']
    cols_name = [ x for x in df_csv.columns if x not in cols_drop]
    unique_step = df_csv.select(cols_name).distinct().cache()

    unique_step.count()

    table_name= 'unique_step'

    unique_step.createOrReplaceTempView(table_name)

    cols_name = ",".join([ x for x in cols_name if x not in ['date']])

    #%%

    sql = f"""SELECT {cols_name}, COLLECT_SET(date)  AS date_lst 
            FROM  {table_name}
            GROUP BY {cols_name}
    """
    print(sql)
    tmp = spark.sql(sql)
    #%%
    tmp = tmp.withColumn('date_count', F.size(tmp.date_lst))
    tmp.show()
    tmp = tmp.where('date_count>48').sort('date_count').toPandas()


    return tmp



# %%
df.show()

# Count the step frequency of each OD in each period
trips_freq = cal_step_freq(df)

# %%
#! 识别 period 1、2 的桥上记录

bridge_map = {1:'Humen bridge', 2:'Nanshan bridge', 3:'HZMB'}

# bridge congestion
def extract_bridge_from_instruction(df):
    view_name = 'df_bridge'
    df.createOrReplaceTempView(view_name)
    sql = f"""SELECT *, (
                CASE 
                    WHEN instruction LIKE "%虎门大桥%" then 1
                    WHEN instruction LIKE "%南沙大桥%" then 2
                    WHEN instruction LIKE "%港珠澳大桥%" then 3
                ELSE -1 END) AS bridge
            FROM {view_name}
        """
    df = spark.sql(sql).where('bridge > 0 ')
    
    return df


def agg_speed_for_bridge(df):
    view_name = 'df_bridge'
    df.createOrReplaceTempView(view_name)
    sql = f"""SELECT period, timestamp, bridge, AVG(speed) as avg_speed
            FROM {view_name}
            GROUP BY period, timestamp, bridge
            ORDER BY period, timestamp, bridge
    """
    
    return spark.sql(sql)


def add_congestion_index(df):
    df.loc[:, 'bridge'] = df.bridge.apply( lambda x: bridge_map[x])

    df = df.merge(
        df.groupby('bridge')[['avg_speed']].max().rename(columns={"avg_speed":'max_speed'}).reset_index(),
        on='bridge'    
    )
    df.loc[:, 'ci'] = df.max_speed / df.avg_speed
    
    return df


# initialize
def df_initialize_for_p1(df_csv):
    df = df_csv.where("date < '2020-01-01' ")\
                .withColumn('t', F.substring(F.col('tripID'), -7, 4))\
                .withColumn('timestamp', F.to_timestamp( F.concat( F.col('date').cast(StringType()), F.lit(' '), F.col('t')) , 'yyyy-MM-dd HHmm'))\
                .withColumn('OD', F.substring(F.col('tripID'), -3, 3).cast(IntegerType()))\
                .withColumn('t', F.hour(F.col('timestamp')) + F.minute(F.col('timestamp'))/F.lit(60) )\
                .withColumn('instruction', F.trim(F.col('instruction')))\
                .withColumn('speed', F.col('distance')/F.col('duration')*F.lit(3.6))
    
    return df


def special_time_filter(data_bridge):
    data_bridge.loc[:, 't'] = data_bridge.timestamp.apply(lambda x: x.hour + x.minute/60)
    time_filter = []
    cur=.25
    while cur < 7:
        time_filter.append(cur)
        cur += .5

    data_bridge.query( 't not in @time_filter ', inplace=True )

    return data_bridge


p1 = df_initialize_for_p1(df_csv).cache()
p3 = df.withColumn('speed', F.col('distance') / F.col('duration') * F.lit(3.6)).drop('period')

p1_bridge = df_pipline(p1, [split_period, extract_bridge_from_instruction, agg_speed_for_bridge]).cache()
p3_bridge = df_pipline(p3, [split_period, extract_bridge_from_instruction, agg_speed_for_bridge]).cache()

data_bridge = pd.concat([p1_bridge.toPandas(), p3_bridge.toPandas()])
data_bridge = add_congestion_index(data_bridge)
data_bridge = special_time_filter(data_bridge)


#%%
# ! 绘制拥堵指数分布曲线


def add_ticks_day(x0, x1, ax,step_time=np.timedelta64(1,'D'), label_step=2 ):

    ticks = []
    tmp = x0
    while tmp <= x1:
        ticks.append(tmp)
        tmp += step_time

    try:
        ax.set_xticks(ticks)
        ax.set_xticklabels([ x._date_repr if i % label_step == 0 else '' for i, x in enumerate(ticks) ])
        return True
    except:
        return False


def get_weekends(x0, x1):
    fisrt_weekend_offset = 5 - x0.dayofweek if x0.dayofweek < 5 else 0
    cur = x0 + np.timedelta64(fisrt_weekend_offset,'D')

    ans = []
    while cur < x1:
        end = cur + np.timedelta64(2,'D')
        if end > x1:
            ans.append([cur, x1])
            break
        ans.append([cur, end])
        cur += np.timedelta64(7,'D')   

    return ans


def draw_congestion_index_by_date_p1(data_bridge, plot_params, y_max=4.2, ax=None, save_fn=None):
    if ax is None:
        fig, ax = plt.subplots(figsize=(18, 6))
    plot_params['ax'] = ax

    data_bridge_sub = data_bridge.query( "period <= 2" )
    x0 = data_bridge_sub.timestamp.min()
    x1 = data_bridge_sub.timestamp.max()+np.timedelta64(15,'m')

    # add 
    # data_0406 = data_bridge.query(" '2019-04-07' <= timestamp < '2019-04-08'")
    # data_0406.loc[:, 'timestamp'] = data_0406.timestamp - np.timedelta64(1,'D')
    # sns.lineplot( data=data_0406, linestyle='-.', zorder=9, **plot_params, legend=False)

    sns.lineplot( data=data_bridge_sub.query("timestamp  < '2019-04-06 00:00:00' "), **plot_params, legend=False)
    sns.lineplot( data=data_bridge_sub.query("timestamp >= '2019-04-07 00:00:00' "), **plot_params)

    holiday = [np.datetime64('2019-04-05 00:00:00'), np.datetime64('2019-04-08 00:00:00')]
    ax.fill_between(holiday, 0, y_max, color='red', alpha=0.1, zorder=2, label='Qingming Festival')

    for index, weekend in enumerate(get_weekends(x0, x1)):
        ax.fill_between(weekend, 0, y_max, color='gray', alpha=0.2, zorder=1, **{'label':'Weekends'} if index==0 else {})
    
    ax.vlines(np.datetime64('2019-04-02 12:00:00'), 0 , y_max, colors='red', linestyle='--', lw=1, label='Nansha Bridge opened')

    ax.set_xlim(x0, x1)
    ax.set_ylim(0.90, y_max)
    add_ticks_day(x0, x1, ax)
    ax.legend()
    
    ax.set_xlabel('a) Average daily congestion index by date in period 1 and 2', fig_title_font)
    ax.set_ylabel("Congestion Index", fig_title_font)

    if save_fn:
        plt.savefig(save_fn, dpi=300)

    return ax


def draw_congestion_index_by_date_p3(data_bridge, plot_params, y_max=4.2, ax=None, save_fn=None):
    if ax is None:
        fig, ax = plt.subplots(figsize=(18, 6))
    plot_params['ax'] = ax

    data_bridge_sub = data_bridge.query( "period > 2" )
    # x0 = data_bridge_sub.timestamp.min()+np.timedelta64(45+5*60,'m')
    # x1 = data_bridge_sub.timestamp.max()+np.timedelta64(15,'m')

    x0 = pd.to_datetime('2020-05-07 00:00:00')
    x1 = x0 + np.timedelta64(22, 'D')


    sns.lineplot( data=data_bridge_sub, **plot_params)

    for index, weekend in enumerate(get_weekends(x0, x1)):
        ax.fill_between(weekend, 0, y_max, color='gray', alpha=0.2, zorder=1, **{'label':'Weekends'} if index==0 else {})

    ax.vlines(np.datetime64('2020-05-15 09:00:00'), 0 , y_max, colors='blue', linestyle='--', lw=1, label='Humen Bridge reopened')

    ax.set_ylim(0.90, y_max)
    ax.set_xlim(x0, x1)
    add_ticks_day(x0, x1, ax)
    ax.legend()
    # plt.savefig('ci_dist_p3.jpg', dpi=300)

    ax.set_xlabel('b) Average daily congestion index by date in period 3 and 4', fig_title_font)
    ax.set_ylabel("Congestion Index", fig_title_font)

    
    return ax


def draw_congesion_index_dist(df, ax=None):
    """绘制周期内一天中的分布情况的分布情况

    Args:
        data_bridge ([type]): [description]
        ax ([type], optional): [description]. Defaults to None.
    """
    data_bridge = df.copy()
    if ax is None:
        fig, ax = plt.subplots(figsize=(8, 6))
        
    # sns.lineplot(data=data, x='t', y='ci', hue='period', style='bridge', ax=ax)
    idnx = data_bridge.query( " '2019-04-04 12:00:00' <= timestamp < '2019-04-08 09:00:00'  ").index
    data_bridge.loc[idnx, 'period'] = 5 

    sns.lineplot(data=data_bridge.query('period < 5'), x='t', y='ci', hue='bridge', style='period', size='period', hue_order=['Humen bridge', 'Nanshan bridge', "HZMB"],  sizes=[1.5, 1.5, 2, 0.5], ax=ax)
    ax.set_xlim(0, 24)
    ax.set_xticks(range(0, 28, 4))
    ax.set_xlabel('c) Daily congestion index distribution in the 4 periods', fig_title_font)
    ax.set_ylabel('Congestion Index', fig_title_font)
    ax.legend(loc='upper left')

    return data_bridge


def congesion_index_fig(data_bridge, out_fn='../result/fig/fig_1_congestion_distribution.jpg'):
    plot_params = {"x":'timestamp', 
                    "y":'ci', 
                    "hue":'bridge',
                    "hue_order": ['Humen bridge', 'Nanshan bridge', "HZMB"],
                    "zorder": 9,
                    "alpha": .9
                }

    fig, ax = plt.subplots(3, figsize=(15, 4*3))
    draw_congestion_index_by_date_p1(data_bridge, plot_params, ax=ax[0])
    draw_congestion_index_by_date_p3(data_bridge, plot_params, ax=ax[1])
    draw_congesion_index_dist(data_bridge, ax=ax[2])

    plt.tight_layout(h_pad=1.5)
    
    if out_fn is not None:
        fig.savefig(out_fn, dpi=300, pad_inches=0, bbox_inches='tight')

    return fig

congesion_index_fig(data_bridge)

# %%
